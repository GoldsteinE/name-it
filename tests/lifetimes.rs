use futures::{pin_mut, FutureExt as _};
use name_it::name_it;

async fn bar() {}

#[name_it(Test)]
async fn has_lifetime(x: &str) -> String {
    bar().await;
    x.to_string()
}

#[test]
fn should_work() {
    let x = String::from("Hi!");
    let y: Test = has_lifetime(x.as_str());
    pin_mut!(y);
    assert_eq!(y.now_or_never().unwrap(), "Hi!");
}
