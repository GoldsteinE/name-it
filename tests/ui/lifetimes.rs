use futures::{pin_mut, FutureExt as _};

use name_it::name_it;

async fn foo() {}

#[name_it(Test)]
async fn has_lifetime(x: &str) -> String {
    foo().await;
    x.to_string()
}

fn main() {
    let x = String::from("test");
    let y = has_lifetime(x.as_str());
    pin_mut!(y);
    drop(x);
    y.now_or_never().unwrap();
}
