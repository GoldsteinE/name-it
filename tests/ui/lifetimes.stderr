error[E0505]: cannot move out of `x` because it is borrowed
  --> tests/ui/lifetimes.rs:17:10
   |
15 |     let y = has_lifetime(x.as_str());
   |                          ---------- borrow of `x` occurs here
16 |     pin_mut!(y);
17 |     drop(x);
   |          ^ move out of `x` occurs here
18 |     y.now_or_never().unwrap();
19 | }
   | - borrow might be used here, when `y` is dropped and runs the destructor for type `Named<_has_lifetime_impl::Test<'_>>`
